src/Infrastructure/BotSharp.MCP/Hooks/MCPToolAgentHook.cs

```diff
@@ -0,0 +1,68 @@
+using BotSharp.Abstraction.Agents;
+using BotSharp.Abstraction.Agents.Enums;
+using BotSharp.Abstraction.Agents.Models;
+using BotSharp.Abstraction.Agents.Settings;
+using BotSharp.Abstraction.Conversations;
+using BotSharp.Abstraction.Functions.Models;
+using BotSharp.Core.Mcp;
+using BotSharp.Core.MCP;
+using Microsoft.Extensions.DependencyInjection;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace BotSharp.MCP.Hooks;
+
+public class MCPToolAgentHook : AgentHookBase
+{
+    public override string SelfId => string.Empty;
+
+    public MCPToolAgentHook(IServiceProvider services, AgentSettings settings)
+        : base(services, settings)
+    {
+    }
+
+    public override void OnAgentMCPToolLoaded(Agent agent)
+    {
+        if (agent.Type == AgentType.Routing)
+            return;
+        var conv = _services.GetRequiredService<IConversationService>();
+        var isConvMode = conv.IsConversationMode();
+        if (!isConvMode) return;
+
+        agent.SecondaryFunctions ??= [];
+
+        var functions = GetMCPContent(agent);
+
+        foreach (var fn in functions)
+        {
+            if (!agent.SecondaryFunctions.Any(x => x.Name.Equals(fn.Name, StringComparison.OrdinalIgnoreCase)))
+            {
+                agent.SecondaryFunctions.Add(fn);
+            }
+        }
+    }
+
+    private IEnumerable<FunctionDef> GetMCPContent(Agent agent)
+    {
+        List<FunctionDef> functionDefs = new List<FunctionDef>();
+        var mcpClientManager = _services.GetRequiredService<MCPClientManager>();
+        var mcps = agent.McpTools;
+        foreach (var item in mcps)
+        {
+            var ua = mcpClientManager.Factory.GetClientAsync(item.ServerId).Result;
```

Comment:
Two async=>sync here, suggest having a async method and do async to sync only in one place